File: app/src/main/java/se/gustavkarlsson/skylight/android/dagger/components/ApplicationComponent.java
package se.gustavkarlsson.skylight.android.dagger.components;

import org.threeten.bp.Clock;

import javax.inject.Singleton;

import dagger.Component;
import se.gustavkarlsson.skylight.android.background.UpdateJob;
import se.gustavkarlsson.skylight.android.background.UpdateScheduler;
import se.gustavkarlsson.skylight.android.background.providers.AuroraReportProvider;
import se.gustavkarlsson.skylight.android.dagger.modules.definitive.ActivityModule;
import se.gustavkarlsson.skylight.android.dagger.modules.definitive.EvaluationModule;
import se.gustavkarlsson.skylight.android.dagger.modules.definitive.LatestAuroraReportObservableModule;
import se.gustavkarlsson.skylight.android.dagger.modules.replaceable.AuroraReportModule;
import se.gustavkarlsson.skylight.android.dagger.modules.replaceable.BackgroundUpdateTimeoutModule;
import se.gustavkarlsson.skylight.android.dagger.modules.replaceable.ClockModule;
import se.gustavkarlsson.skylight.android.dagger.modules.replaceable.ForegroundUpdateTimeoutModule;
import se.gustavkarlsson.skylight.android.dagger.modules.replaceable.LatestAuroraReportCacheModule;
import se.gustavkarlsson.skylight.android.dagger.modules.replaceable.UpdateSchedulerModule;

@Component(modules = {
		LatestAuroraReportCacheModule.class,
		EvaluationModule.class,
		AuroraReportModule.class,
		ClockModule.class,
		LatestAuroraReportObservableModule.class,
		BackgroundUpdateTimeoutModule.class,
		ForegroundUpdateTimeoutModule.class,
		UpdateSchedulerModule.class
})
@Singleton
@SuppressWarnings("WeakerAccess")
public interface ApplicationComponent {
	AuroraReportProvider getAuroraReportProvider();
	UpdateJob getUpdateJob();
	UpdateScheduler getUpdateScheduler();
	Clock getClock();

	MainActivityComponent getMainActivityComponent(ActivityModule activityModule);
}

File: app/src/main/kotlin/se/gustavkarlsson/skylight/android/dagger/components/ApplicationComponent.kt
package se.gustavkarlsson.skylight.android.dagger.components

import dagger.Component
import se.gustavkarlsson.skylight.android.dagger.modules.*
import se.gustavkarlsson.skylight.android.gui.activities.main.MainViewModelFactory
import se.gustavkarlsson.skylight.android.gui.activities.main.fragments.aurora_chance.AuroraChanceViewModelFactory
import se.gustavkarlsson.skylight.android.gui.activities.main.fragments.aurora_factors.AuroraFactorsViewModelFactory
import se.gustavkarlsson.skylight.android.services.Scheduler
import se.gustavkarlsson.skylight.android.services.Settings
import se.gustavkarlsson.skylight.android.services_impl.scheduling.UpdateJob
import javax.inject.Singleton

@Component(modules = arrayOf(
	ContextModule::class,
	TimeModule::class,
	SystemServiceModule::class,
	SharedPreferencesModule::class,
	SettingsModule::class,
	NotifierModule::class,
	UpdateSchedulerModule::class,
	EvaluationModule::class,
	AuroraReportModule::class,
	CacheModule::class,
	LocationModule::class,
	AuroraFactorsModule::class,
	DarknessModule::class,
	VisibilityModule::class,
	GeomagLocationModule::class,
	KpIndexModule::class,
	LocationNameProviderModule::class,
	ViewModelsModule::class,
	FormattingModule::class,
	LocalizationModule::class
))
@Singleton
interface ApplicationComponent {
	fun getUpdateJob(): UpdateJob

	fun getAuroraChanceViewModelFactory(): AuroraChanceViewModelFactory
	fun getAuroraFactorsViewModelFactory(): AuroraFactorsViewModelFactory
	fun getMainViewModelFactory(): MainViewModelFactory
	fun getSettings(): Settings
	fun getScheduler(): Scheduler
}
